src/tools/rustfmt/src/parse/session.rs

```diff
@@ -230,17 +230,6 @@ impl ParseSess {
         }
     }
 
-    pub(crate) fn emit_or_cancel_diagnostic(&self, diagnostic: &mut Diagnostic) {
-        self.parse_sess.span_diagnostic.emit_diagnostic(diagnostic);
-        // The Handler will check whether the diagnostic should be emitted
-        // based on the user's rustfmt configuration and the originating file
-        // that caused the parser error. If the Handler determined it should skip
-        // emission then we need to ensure the diagnostic is cancelled.
-        if !diagnostic.cancelled() {
-            diagnostic.cancel();
-        }
-    }
```

Comment:
Admittedly I've only done a very cursory pass through the rustfmt portions of the changes, but just wanted to note this diff gave me pause as it seems like it could introduce an undesirable change in rustfmt behavior for which we don't have test coverage IIRC.

However, even if I'm correct that shouldn't be viewed as a blocker because I've been meaning to do some cleanup on rustfmt's mod parsing and resolution that would obviate the need for what we were doing here to begin with.

Reply:
There's literally no change here. `Diagnostic` has no concept of cancellation, it's just data lying around. The `.cancel()` call only makes sense if you pass the same diagnostic to `emit_diagnostic` *afterwards*, but here it was just dropped (in the caller).

I believe rustdoc had similar non-behavior. You may want an audit for this confusion in semantics, there may be a rustfmt bug that assumed these diagnostics had to be cancelled.

Reply:
I'll try to take a deeper look at the rustfmt side over the weekend, but my recollection is that it was something we did because not cancelling the diagnostic was triggering an ice complaining about errors being constructed but not emitted on some edge cases we were trying to handle

Reply:
Yeah, that's `DiagnosticBuilder`, the type that's the focus of this PR.

This, OTOH, is a `Diagnostic`, the "data" in the `DiagnosticBuilder`, but in this case there's no `DiagnosticBuilder` around it, you just get a `Vec<Diagnostic>` from the rest of the system.

If that was `Vec<DiagnosticBuilder>` at any point in the past, you would've *theoretically* had to worry about cancellation, but you would've used `db.emit();` instead of `handler.emit_diagnostic(&db); db.cancel();` (and `.emit()` would do the cancellation for you).

Reply:
Gotcha :+1:

We originally implemented all this back in the days when the parser was still baked into the now-extinct libsyntax mod and in order to deal with any potential parser errors originating in [ignored](https://rust-lang.github.io/rustfmt/?version=v1.4.38&#ignore) files had this hack to intercept and either emit or suppress/cancel to not display the error + prevent the aforementioned ice.

I was thinking we were potentially going to need to make some more changes on our side in rustfmt first in order for it to continue to play nicely, but fortunately, you're right and that won't be necessary/impactful here